### PR TITLE
Handle legacy passwords

### PR DESCRIPTION
**Pull request by andysbolton**
This PR creates a new column called 'UserPasswordMd5' on the User table and adds a function that will compare a user's plaintext password against their legacy password's MD5 hash and bcrypt it if it exists and matches.

### CODE REVIEW

File: database/database_test.go [hunk 1]
```diff
@@ -509,3 +511,112 @@ func TestCreateUser(t *testing.T) {
...
+	err = d.HandlePasswordMigration(&user, &credentials)
+
+	if err != nil && err == ErrWrongPassword {
+		t.Log("MD5 hashes do not match")
+	} else {
+		t.Fatal("MD5 hashes match")
+	}
+}
+
+type AnyByte struct{}
```
Comment (DarthHater) [inline database/database_test.go]: I would move this to the top of the file (more golangy, I guess)

File: main.go [hunk 2]
```diff
@@ -531,12 +530,14 @@ func checkCredentials(c *gin.Context, d database.IDatabase) {
 
 	err = bcrypt.CompareHashAndPassword(user.Password, []byte(credentials.Password))
 	if err != nil {
-		log.WithFields(log.Fields{"username": user.Username}).Error("Wrong password")
-		c.JSON(http.StatusUnauthorized, gin.H{"err": "Wrong password"})
-		return
+		if err = d.HandlePasswordMigration(&user, &credentials); err != nil {
```
Comment (DarthHater) [inline main.go]: Someone learned how to shorthand!

File: model/user.go [hunk 3]
```diff
@@ -10,6 +12,7 @@ type User struct {
 	EmailAddress string
 	Password []byte
 	UserRole      int
+	UserPasswordMd5	sql.NullString
```
Comment (DarthHater) [inline model/user.go]: You can just do string here, I think? Any reason to use sql.NullString?

File: model/user.go [hunk 4]
```diff
@@ -26,3 +29,8 @@ func (u *User) HashPassword(password string) (err error) {
 
 	return nil
 }
+
+func (u *User) HashPasswordMd5(password string) (hash [16]byte) {
```
Comment (DarthHater) [inline model/user.go]: Very nice!

File: model/user.go (hunk 3 above)
Comment (andysbolton) [reply model/user.go]: I seemed to be getting an error here when scanning null into that field, but I'll play around with it a little bit. 

https://github.com/DarthHater/bored-board-service/blob/HandleLegacyPasswords/database/database.go#L143

**Pull request merged**